src/cmd/ksh93/sh/fault.c

```diff
@@ -50,6 +55,38 @@ struct Siginfo
 static char	indone;
 static int	cursig = -1;
 
+/* Write a primitive backtrace to stderr. */
+void dump_backtrace(int max_frames, int skip_levels) {
+      void *callstack[128];
+      const int n_max_frames = sizeof(callstack) / sizeof(callstack[0]);
+      int n_frames = backtrace(callstack, n_max_frames);
+      char **symbols = backtrace_symbols(callstack, n_frames);
+
+      int i;
+      for (i = skip_levels; i < n_frames; i++) {
+	 Dl_info info;
+	 if (dladdr(callstack[i], &info) && info.dli_sname) {
+	       fprintf(stderr, "%-3d %s + %td\n", i - skip_levels,
```

Comment:
Can we use ```write()``` here ?  ```printf()``` is not signal safe.

Reply:
The problem is how to get formatted output. Even `snprintf()` isn't async safe. However, it appears that in practice `snprintf()` is safe if you don't use `%f` (which is known to call `malloc()` in some implementations) or other advanced formatting options. Also, the other functions being used, e.g., `dladdr()`, aren't async safe. 

We could forego the formatting and just write the `info.dli_sname` string. But I'm hoping to find a reasonably portable way to determine line numbers from the byte offset currently displayed.

The benefits of displaying this info far outweigh the risk that any of these calls will interact badly with the context at the time of the fault. But I do agree we should eliminate the `fprintf()` since performing stdio is likely to be the function most likely to interact badly with the interrupted context.